### PR TITLE
Refactor configdata.pm to be generated by template

### PR DESCRIPTION
**Pull request by levitte**
To have the `configdata.pm` text embedded in Configure was kind of ugly,
and becomes clearer if put into a template file, `configdata.pm.in`.

First, we need to make a module `OpenSSL::Template` from the central parts
of `util/dofile.pl`, and also reduce the amount of ugly code with more proper
use of `Text::Template`.  `OpenSSL::Template` is a simple subclass of
`Text::Template`.

Now we can use `OpenSSL::Template` to generate `configdata.pm` from
`configdata.pm.in`.

We also modify `configdata.pm` to be the build file generator, and run
it from Configure.  The benefit with that is that developers who
tinker and play with the build file can do a "factory reset" without
having to go through the configuration process, i.e. they can re-use
the config data the already have.

### CODE REVIEW

File: Configure [hunk 1]
```diff
@@ -2386,452 +2387,49 @@ foreach (grep /_(asm|aux)_src$/, keys %target) {
...
-            print OUT " "x4, quotify("perl", $_), " => ", quotify("perl", $unified_info{$_}), ",\n";
-        } else {
-            print OUT " "x4, quotify("perl", $_), " =>\n";
-            $recurse->(8, $unified_info{$_});
-        }
-    }
-    print OUT <<"EOF";
+my $configdata_outname = 'configdata.pm';
+print "Creating $configdata_outname\n";
+open CONFIGDATA, ">$configdata_outname"
```
Comment (levitte) [inline Configure]: Note to self: modify to create configdata.pm.new and rename to configdata.pm on success only

File: configdata.pm.in [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+ sub out_item {
+     my $ref = shift;
+     # Available options:
+     # indent           => callers indentation (int)
+     # delimiters       => 1 if outer delimiters should be added
+     my %opts = @_;
+
+     my $indent = $opts{indent} // 0;
+     # Indentation of the whole structure, where applicable
+     my $nlindent1 = defined $indent ? "\n" . ' ' x $indent : ' ';
```
Comment (richsalz) [inline configdata.pm.in]: I'm assuming this works, but I'm confused why.  Doesn't line 11 guarantee that $indent is defined?  Should that be a test against zero?  If I'm wrong, can you explain to this noob why?

File: configdata.pm.in [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+     # Indentation of individual items, where applicable
+     my $nlindent2 = defined $indent ? "\n" . ' ' x ($indent + 4) : ' ';
+
+     my $product;      # Finished product, or reference to a function that
+                       # produces a string, given $_
+     # The following are only used when $product is a function reference
+     my $delim_l;      # Left delimiter of structure
+     my $delim_r;      # Right delimiter of structure
+     my $delim_b;      # "Between" delimiter, between items
```
Comment (richsalz) [inline configdata.pm.in]: not $separator ?

**Comment by richsalz** [comment]
Moving past the "fear of change" factor, this seems like a good idea.  Should configdata get a new name, like buildfiles or something?  And is it still right to have it be a ".pm" file?

Thinking further, what is the difference now between Configure and config?  Should a future PR merge them?

File: configdata.pm.in (hunk 2 above)
Comment (levitte) [reply configdata.pm.in]: Old code from when it might be undefined, so obvious cleanup needed. Thanks

**Comment by levitte** [comment]
> And is it still right to have it be a ".pm" file?

Yes. It's primarily a module, and if you `git grep 'use *configdata'`, you'll see that for yourself. It's function as a script is a late addition (came in 1.1.1)

> Thinking further, what is the difference now between Configure and config? Should a future PR merge them?

The difference is that `config` figures out the appropriate config target, then calls `Configure`. I agree they should become one, and have had that on my mind for a while, but that's a helluva job...

**Comment by richsalz** [comment]
I looked at config after writing my comment, blanched, and then walked away. :) I still might take a crack at perlizing it.  Thanks for the .pm/script clarification.

File: configdata.pm.in (hunk 2 above)
Comment (levitte) [reply configdata.pm.in]: Done

File: configdata.pm.in (hunk 3 above)
Comment (levitte) [reply configdata.pm.in]: Sure

File: Configure (hunk 1 above)
Comment (levitte) [reply Configure]: Done

**Comment by levitte** [comment]
Ping

**Comment by levitte** [comment]
Ping

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by mattcaswell — APPROVED** [review]
Just one nit. Approved assuming that's fixed.

**Pull request closed without merging**